Implemented new scripts for update 4.1

trouvant opened the pull request:
Resolves #9.

Note that I chose not to implement _pull(), for, unless I'm mistaken, it wouldn't work well with GML in its present form. The problem stems from the lack of support for resizing an array except to expand it, and without any way to directly access the memory on the heap associated with an array, it appears to me impossible to contract an array to the appropriate size once elements are removed. One should use _without() or, preferably, ds_lists for similar behaviour.

trouvant committed:
Implemented new scripts for update 4.1

twisterghost commented:
Good stuff. I'll review this soon.

thijsmie commented:
This is looking really good! I did spot a few points of possible optimization, I will properly test, review and report on these over the weekend.

twisterghost commented:
Looks like the scripts are _in_ the repo, but haven't been integrated into the project itself, so it doesn't show up in the editor. Do you mind pushing a commit with them all added to the editor?

twisterghost commented on src/scripts/_chunk/_chunk.gml:
```diff
@@ -0,0 +1,30 @@
+
+var i;
+var _len;
+var __chunk;
```
the leading and underscore and double underscore naming convention is new to this repo. Is that to avoid variable collisions? Scoping in scripts will fix that problem - if you `var` to initialize a variable in a script, it overwrites any instance variable in the scope of the script without affecting the instance.

If not for that purpose, I'd like to hear the reasoning. I'm open to conventions, but trying to keep code style in the repo consistent.

twisterghost commented on src/scripts/_chunk/_chunk.gml:
```diff
@@ -0,0 +1,30 @@
+_len = array_length_1d(argument0);
+
+for (i = 0; i < _len; i++) {
+    __chunk[i div argument1, i mod argument1] = argument0[i];
```
For code style, prefer using the arithmetic characters over their written form. This should be:

`__chunk[i / argument1, i % argument1] = argument0[i];`

thijsmie replied:
Also, the loop should go from _len-1 to 0 instead to avoid continuous resizing of __chunk.

twisterghost commented on src/scripts/_difference_by/_difference_by.gml:
```diff
@@ -0,0 +1,59 @@
+// => [1];
+*/
+
+if (argument_count == 0) return array_create(0);
```
You can just return `[0]` instead of using `array_create`. GM:S2 GML has support for native array definitions.

thijsmie replied:
array_create(0) is an array of zero size, so return [] or array_create(0) are both fine I think.

twisterghost commented:
Looking really good. I'd love to see what @thijsmie finds when reviewing as well, and would be happy to help write tests for these scripts - certainly need them to ensure everything is in order.

trouvant commented:
I've made the changes requested (including actually adding the scripts to the project--oops), made sure that I've fixed any documentation mistakes, and I've begun writing the tests. I should be finished before the weekend is out. 

One question, though: I implemented a version of _reverse() before realizing that one had already been added to the library. However, mine mutates the array passed to the script, à la Lodash, but the one that has already been implemented creates a new array. Do you want to keep the one already there, or replace the mutative function? Or, alternatively, should we keep both and rename one of them?

twisterghost commented:
For now, I'm open to the idea of keeping both and renaming yours `_reverse_mut` and specifying that it mutates input. That can keep this to be a minor release rather than a breaking release, and in the future we can assess if we'd like to combine them in a major release.
